src/v2/controllers/vreplicaset_controller/trusted/rely_guarantee.rs

```diff
@@ -0,0 +1,167 @@
+        req.obj.kind == Kind::PodKind ==> 
+            req.obj.metadata.resource_version.is_Some()
+            && !{
+                let etcd_obj = s.resources()[req.key()];
+                let owner_references = etcd_obj.metadata.owner_references.get_Some_0();
+                &&& s.resources().contains_key(req.key())
+                &&& etcd_obj.metadata.resource_version.is_Some()
+                &&& etcd_obj.metadata.resource_version == req.obj.metadata.resource_version
+                &&& etcd_obj.metadata.owner_references.is_Some()
+                &&& exists |vrs: VReplicaSetView| 
+                    #[trigger] owner_references.contains(vrs.controller_owner_ref())
+            }
+    }
+}
+
+// All delete requests to pods carry a precondition on the resource version,
+// and other controllers don't try to delete pods owned by a VReplicaSet.
+pub open spec fn vrs_rely_delete_req(req: DeleteRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        req.key.kind == Kind::PodKind ==>
+            req.preconditions.is_Some()
+            && req.preconditions.get_Some_0().resource_version.is_Some()
+            && !{
+                let etcd_obj = s.resources()[req.key];
+                let owner_references = etcd_obj.metadata.owner_references.get_Some_0();
+                &&& s.resources().contains_key(req.key)
+                &&& etcd_obj.metadata.resource_version.is_Some()
+                &&& etcd_obj.metadata.resource_version
+                    == req.preconditions.get_Some_0().resource_version
+                &&& etcd_obj.metadata.owner_references.is_Some()
+                &&& exists |vrs: VReplicaSetView| 
+                    #[trigger] owner_references.contains(vrs.controller_owner_ref())
+            }
+    }
+}
+
+pub open spec fn vrs_rely(other_id: int) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        forall |msg| {
+            &&& #[trigger] s.in_flight().contains(msg)
+            &&& msg.content.is_APIRequest()
+            &&& msg.src == HostId::Controller(other_id)
+        } ==> match msg.content.get_APIRequest_0() {
+            APIRequest::CreateRequest(req) => vrs_rely_create_req(req)(s),
+            APIRequest::UpdateRequest(req) => vrs_rely_update_req(req)(s),
+            APIRequest::UpdateStatusRequest(req) => vrs_rely_update_status_req(req)(s),
+            APIRequest::DeleteRequest(req) => vrs_rely_delete_req(req)(s),
+            _ => true,
+        }
+    }
+}
+
+// VRS Guarantee Condition
+
+// VRS only creates pods owned by a VReplicaSet.
+pub open spec fn vrs_guarantee_create_req(req: CreateRequest) -> StatePred<ClusterHistory> {
+    |h: ClusterHistory| {
+        let owner_references = req.obj.metadata.owner_references.get_Some_0();
+        &&& req.obj.kind == Kind::PodKind
+        &&& req.obj.metadata.owner_references.is_Some()
+        &&& exists |vrs: VReplicaSetView| 
+            #[trigger] owner_references.contains(vrs.controller_owner_ref())
```

Comment:
@codyjrivera You might want to enhance it to state that there is only one controller_owner_ref in owner_references which is vrs.

Reply:
this is guaranteed by an init invariant.

Reply:
Yes, `each_object_in_etcd_has_at_most_one_controller_owner` can do the job, but then you will not be able to prove that VRS's guarantee implies another controller's rely --- you can only prove that with VRS's guarantee condition and this invariant, VRS's creation request will not interfere with another controller.

Ideally, you could structure the compositional proof in a way that (1) first prove A's guarantee implies B's rely, and (2) then prove B's liveness under B's rely.

The proof of (2) will be much more complex than (1), so you only want to do it once and it should not depend on A. The proof of (1) should be much easier and you will need to do it for each pair of controllers.